ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateArchive.java

```diff
@@ -169,34 +177,41 @@ public boolean isWorldStateAvailable(final Hash rootHash, final Hash blockHash)
       return Optional.of(persistedState);
     } else {
       try {
-        BlockHeader persistedHeader = blockchain.getBlockHeader(persistedState.blockHash()).get();
-        BlockHeader targetHeader = blockchain.getBlockHeader(blockHash).get();
+        Optional<BlockHeader> persistedHeader =
+            blockchain.getBlockHeader(persistedState.blockHash());
 
         final List<TrieLogLayer> rollBacks = new ArrayList<>();
         final List<TrieLogLayer> rollForwards = new ArrayList<>();
+        if (persistedHeader.isEmpty()) {
+          getTrieLogLayer(persistedState.blockHash()).ifPresent(rollBacks::add);
+        } else {
+          BlockHeader targetHeader = blockchain.getBlockHeader(blockHash).get();
```

Comment:
non-blocking: it would help with readability to resolve the Optional persistedHeader once here and use it in the rest of the block.  all of the naked `Optional.get()` make me nervous (despite being safe here).  If we had a small refactor of the flow, those get's could be landmines

Reply:
ok I will do that

Reply:
Done

Reply:
Done